File: spring-jpa/querydsl/src/main/java/study/querydsl/service/MemberService.java
package study.querydsl.service;


import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import study.querydsl.dto.MemberTeamDto;
import study.querydsl.repository.MemberRepository;

import java.util.List;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class MemberService {

    private final MemberRepository memberRepository;

    public List<MemberTeamDto> search(String username, Integer age) {
        return memberRepository.search(username, age);
    }

    public Page<MemberTeamDto> searchPageSimple(String username, Integer age, Pageable pageable) {
        return memberRepository.searchPageSimple(username, age, pageable);
    }

    public Page<MemberTeamDto> searchPageComplex(String username, Integer age, Pageable pageable) {
        return memberRepository.searchPageComplex(username, age, pageable);
    }
}

File: spring_coroutines_training/src/main/kotlin/app/training/service/MemberService.kt
package app.training.service

import app.training.dto.MemberDto
import app.training.repository.MemberRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class MemberService(
    private val memberRepository: MemberRepository,
) {
    @Transactional(readOnly = true)
    suspend fun findMyInfo(id: Long): MemberDto {
        return withContext(Dispatchers.IO) {
            println("MemberService.findMyInfo 내부 : ${Thread.currentThread().name}")
            val member = memberRepository.findByIdOrNull(id) !!
            delay(3000)

            MemberDto(
                id = member.id !!,
                name = member.name,
            )
        }
    }
}